geps/gep-1897.md

```diff
@@ -0,0 +1,113 @@
+# GEP-1897: TLS from Gateway to Backend for ingress
+
+* Issue: [#1897](https://github.com/kubernetes-sigs/gateway-api/issues/1897)
+* Status: Provisional
+
+## TLDR
+
+This document specifically addresses the topic of conveying HTTPS from the Gateway
+dataplane to the backend (backend TLS termination), and intends to satisfy the single
+use case “As a client implementation of Gateway API, I need to know how to connect to
+a backend pod that has its own certificate”. TLS configuration can be a nebulous topic,
+so in order to drive resolution this GEP focuses only on this single piece of functionality.
+
+## Goals
+
+1. The solution should satisfy the following use case: the backend pod has its own
+certificate and the gateway implementation client needs to know how to connect to the
+backend pod. (Use case #4 in [Gateway API TLS Use Cases](#references))
+2. In terms of the Gateway API personas, the application developer should control the
+gateway to backend TLS settings, not the cluster operator, as requiring a cluster operator
+to manage certificate renewals and revocations would be extremely cumbersome.
+3. The solution should consider client certificate settings used in the TLS handshake **from
+Gateway to backend**, such as TLS versions and cipher suites. (Use case #5 in [Gateway API TLS Use Cases](#references))
+
+## Non-Goals
+
+1. Changes to the existing mechanisms for edge or passthrough TLS termination
+2. Providing a mechanism to decorate multiple route instances
+3. Service mesh use cases
+4. Mutual TLS use cases
+5. TLSRoute, TCPRoute, or UDPRoute use cases
```

Comment:
While I acknowledge that it is critical to keep the scope reduced in order to make forward progress, I am a bit concerned with making these "Non-goals" (these 2 in particular, but to some extent egress and mesh as well).

I worry that by only considering the narrowest case, we will pigeon-hole ourselves into a non-flexible API. I don't want us to implement "TLSSettingsForHTTPFromIngress" (logically or literally) now, and realize we actually wanted it to be a more general "TLSSettingsFromIngress" (support TCPRoute) or "TLSSettings" (support non-ingress) cases and not be able to move.

Its possible that we will just get lucky and this won't happen, but relying on that seems unlikely to work out well.

I would feel more comfortable if the GEP was more like https://github.com/kubernetes-sigs/gateway-api/pull/1643, which I view as the golden standard for a "What and why" GEP, which explores the problem space and prior art.

Without this, I worry we are just going to solve one very very narrow case and create an awkward long term API.

I would also like to remind folks that the easiest way to experiment with these type of customizations is to _just go implement it_. Gateway API is designed around extensibility; there is nothing stopping a custom implementation from implementing TLS origination today. This serves as an incredibly valuable reference point for discussion - at the end of the day, we get far more info from a real implementation than debating on a doc. That isn't to say that this isn't valuable in core - it absolutely is. But its important that we do it right in core than that we do it fast (or even at all!).

Reply:
As this comment stands there's a lot of room for interpretation and there's some ambiguity in the suggestions for action. In this state I could see it hanging around for some time without being resolved, so I want to try and get it distilled. The "just go implement it" suggestion is not something we the maintainers feel we should be pushing for in this case, but for everything else we would recommend creating prescriptive and explicit suggestions (ideally with Github's review suggestions feature when feasible, as these are clear and actionable). Are you OK with resolving this comment in favor of reworking it into more explicit suggestions for change?